File: rust/src/node.rs
use serde_json::Value;

// use crate::hash::HashMap;
use crate::string::str;
use crate::actions::actor::ActorActions;
use crate::actions::component::ComponentActions;
use crate::actions::core::CoreActions;
use crate::actions::entity::EntityActions;
use crate::actions::symbol::SymbolActions;
use crate::context::Context;
use crate::options::Options;
// use crate::storage::Storage;
use crate::types::{
    Actions as ActionsObject,
    // ActorsPayload,
    // ComponentsPayload,
    // ComponentPayload,
    // ComponentTarget,
    // EntityPayload,
    // InputPayload,
    // SymbolPayload,
};

// TODO: replace this with options.isAuthority

/**
 * The actions object combines all the actions from different modules.
 */
pub struct Actions {}

impl ActorActions for Actions {}

impl ComponentActions for Actions {}

impl CoreActions for Actions {}

impl EntityActions for Actions {}

impl SymbolActions for Actions {}

pub fn get_actions() -> Actions {
    Actions {}
}

fn setup_actions<'a>(actions: &mut ActionsObject) -> &'a ActionsObject<'a> {
    actions.insert(
        str("actorInput"),
        Box::new(&(Actions::actor_input as fn(&Value, &Context, &Options))),
    );
    actions.insert(
        str("actors"),
        Box::new(&(Actions::actors as fn(&Value, &Context, &Options)))
    );
    actions.insert(
        str("mergeActors"),
        Box::new(&(Actions::merge_actors as fn(&Value, &Context, &Options))),
    );
    actions.insert(
        str("removeActor"),
        Box::new(&(Actions::remove_actor as fn(&Value, &Context, &Options))),
    );
    actions.insert(
        str("spawnActor"),
        Box::new(&(Actions::spawn_actor as fn(&Value, &Context, &Options)))
    );

    actions.insert(
        str("changeComponent"),
        Box::new(&(Actions::change_component as fn(&Value, &Context, &Options)))
    );
    actions.insert(
        str("components"),
        Box::new(&(Actions::components as fn(&Value, &Context, &Options)))
    );
    actions.insert(
        str("mergeComponents"),
        Box::new(&(Actions::merge_components as fn(&Value, &Context, &Options)))
    );
    actions.insert(
        str("removeComponent"),
        Box::new(&(Actions::remove_component as fn(&Value, &Context, &Options)))
    );
    actions.insert(
        str("upsertComponent"),
        Box::new(&(Actions::upsert_component as fn(&Value, &Context, &Options)))
    );

    actions.insert(
        &String::from("batch"),
        Box::new(&(Actions::batch as fn(&Value, &Context, &Options)))
    );

    actions.insert(
        &String::from("createEntity"),
        Box::new(&(Actions::create_entity as fn(&Value, &Context, &Options))),
    );
    actions.insert(
        &String::from("entities"),
        Box::new(&(Actions::entities as fn(&Value, &Context, &Options))),
    );
    actions.insert(
        &String::from("mergeEntities"),
        Box::new(&(Actions::merge_entities as fn(&Value, &Context, &Options))),
    );
    actions.insert(
        &String::from("removeEntity"),
        Box::new(&(Actions::remove_entity as fn(&Value, &Context, &Options))),
    );

    actions.insert(
        &String::from("createEntity"),
        Box::new(&(Actions::add_symbol as fn(&Value, &Context, &Options) -> Option<u32>)),
    );
    actions.insert(
        &String::from("fetchSymbol"),
        Box::new(&(Actions::fetch_symbol as fn(&Value, &Context, &Options))),
    );
    actions.insert(
        &String::from("getSymbol"),
        Box::new(&(Actions::get_symbol as fn(&Value, &Context, &Options) -> Option<&'a String>)),
    );
    actions.insert(
        &String::from("mergeSymbols"),
        Box::new(&(Actions::merge_symbols as fn(&Value, &Context, &Options))),
    );
    actions.insert(
        &String::from("symbol"),
        Box::new(&(Actions::symbol as fn(&Value, &Context, &Options))),
    );
    actions.insert(
        &String::from("symbols"),
        Box::new(&(Actions::symbols as fn(&Value, &Context, &Options))),
    );

    &actions
}

/**
 * An object that maps the names of actions to their corresponding methods in the ActorActions struct.
 */
pub const ACTIONS: &ActionsObject = setup_actions(&mut ActionsObject::new());

File: ecmascript/src/node.ts
import actorActions, { ActorActionsFactory } from './actions/actor'
import componentActions, { ComponentActionsFactory } from './actions/component'
import coreActions, { CoreActionsFactory } from './actions/core'
import entityActions, { EntityActionsFactory } from './actions/entity'
import symbolActions, { SymbolActionsFactory } from './actions/symbol'

/**
 * A factory function that creates a new instance of the AllActions class that extends the provided Parent class
 * 
 * @param {any} Parent - The class to be extended by the AllActions class.
 * @returns {any} A new class that extends the provided Parent class and the CoreActions class.
 */
export function AllActionsFactory(Parent: any = Object): any {
    return ActorActionsFactory(
        ComponentActionsFactory(
            CoreActionsFactory(
                EntityActionsFactory(
                    SymbolActionsFactory(Parent)
                )
            )
        )
    )
}

/**
 * Combines all the actions from different modules into a single object.
 */
export const allActions = {
    ...actorActions,
    ...componentActions,
    ...coreActions,
    ...entityActions,
    ...symbolActions
}

/**
 * The NodeActions class provides methods for managing nodes in a context.
 */
export class NodeActions extends AllActionsFactory() { }

const __NodeActions__ = new NodeActions()

/**
 * The actions object combines all the actions from different modules.
 */
export const actions = {
    /**
     * Handles input for a specific actor in the current context.
     */
    actorInput: __NodeActions__.actorInput,

    /**
     * Retrieves actors from the current context.
     */
    actors: __NodeActions__.actors,

    /**
     * Merges actors into the current context.
     */
    mergeActors: __NodeActions__.mergeActors,

    /**
     * Removes an actor from the current context.
     */
    removeActor: __NodeActions__.removeActor,

    /**
     * Spawns a new actor in the current context.
     */
    spawnActor: __NodeActions__.spawnActor,

    /**
     * Changes a component in the current context.
     */
    changeComponent: __NodeActions__.changeComponent,

    /**
     * Retrieves components from the current context.
     */
    components: __NodeActions__.components,

    /**
     * Merges components into the current context.
     */
    mergeComponents: __NodeActions__.mergeComponents,

    /**
     * Removes a component from the current context.
     */
    removeComponent: __NodeActions__.removeComponent,

    /**
     * Updates an existing component or inserts a new one if it doesn't exist in the current context.
     */
    upsertComponent: __NodeActions__.upsertComponent,

    /**
     * Processes a batch of payloads in the current context.
     */
    batch: __NodeActions__.batch,

    /**
     * Creates a new entity in the current context.
     */
    createEntity: __NodeActions__.createEntity,

    /**
     * Retrieves entities from the current context.
     */
    entities: __NodeActions__.entities,

    /**
     * Merges entities into the current context.
     */
    mergeEntities: __NodeActions__.mergeEntities,

    /**
     * Removes an entity from the current context.
     */
    removeEntity: __NodeActions__.removeEntity,

    /**
     * Adds a symbol to the current context.
     */
    addSymbol: __NodeActions__.addSymbol,

    /**
     * Fetches a symbol from the current context.
     */
    fetchSymbol: __NodeActions__.fetchSymbol,

    /**
     * Retrieves a symbol from the current context by its index.
     */
    getSymbol: __NodeActions__.getSymbol,

    /**
     * Merges a symbol into the current context.
     */
    mergeSymbol: __NodeActions__.mergeSymbol,

    /**
     * Merges multiple symbols into the current context.
     */
    mergeSymbols: __NodeActions__.mergeSymbols,

    /**
     * Retrieves a symbol from the current context.
     */
    symbol: __NodeActions__.symbol,

    /**
     * Retrieves all symbols from the current context.
     */
    symbols: __NodeActions__.symbols,
}

export default actions
